graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -812,10 +813,86 @@ def _preprocess_graphql_string(graphql_string):
     return graphql_string + '\n'
 
 
+def _sorted_tuple(unordered_list):
+    """Sort a list and turn it into a tuple.
+
+    Args:
+        unordered_list: list of strings
+
+    Returns:
+        tuple of the sorted input list
+    """
+    return tuple(sorted(unordered_list))
+
+
+def _validate_schema_and_ast(schema, ast):
+    """Validate the supplied graphql schema and ast.
+
+    This method wraps around graphql-core's validation to enforce a stricter requirement of the
+    schema-- all directives supported by the compiler must be declared by the schema, regardless of
+    whether each directive is used in the query or not.
+
+    Args:
+        schema: GraphQL schema object, created using the GraphQL library
+        ast: abstract syntax tree representation of a graphql query
+
+    Returns:
+        list containing schema and/or query validation errors
+    """
+    core_graphql_errors = validate(schema, ast)
+
+    # The following directives do not appear in the core-graphql library, but are permissable.
+    whitelisted_directives = set([
+        ('include', ('FIELD', 'FRAGMENT_SPREAD', 'INLINE_FRAGMENT'), ('if',)),
+        ('skip', ('FIELD', 'FRAGMENT_SPREAD', 'INLINE_FRAGMENT'), ('if',)),
+        ('deprecated', ('ENUM_VALUE', 'FIELD_DEFINITION'), ('reason',))
+    ])
+
+    # Extract name, locations and args keys in order to compare schema directives and directives
+    # which are supported by the graphql compiler.
+
+    # Directives provided in the GraphQL string.
+    expected_directives = set(
+        (
+            directive.name,
+            _sorted_tuple(directive.locations),
```

Comment:
Don't we just want these to be sets instead? `_sorted_tuple` seems to just be a workaround for order-independent equality checking, which sets provide.

Reply:
yes, but we can't do a set of sets, so the tuple is kinda a hack. I could change this to a list of sets and change up  how the intersection/disjunction of directives is computed, but I don't have a strong preference either way.